libc-bottom-half/headers/public/wasi/descriptor_table.h

```diff
@@ -0,0 +1,121 @@
+#ifndef DESCRIPTOR_TABLE_H
+#define DESCRIPTOR_TABLE_H
```

Comment:
Are we sure we want to make this into a public API?   So far WASI-libc has not really added new APIs.. and more of the code has been an internal implementation detail.  

If we add new APIs and make them public we then have to start supporting those APIs going forward.  If we can get away with making this internal (at least initially) that would be been IMHO.

Reply:
Yes, I agree that it's a big commitment and not to be taken lightly, and I'm open to postponing making this API public.

I do expect applications will want it, though.  For example, I've updated Rust's `mio` library to support WASI Preview 2's `wasi:io/poll` interface directly rather than go through e.g. POSIX's `poll` or `select`, which reduces the layers of abstraction and their attendant performance and debugging overhead.  Without some way to translate handles into descriptors and vice-versa, that wouldn't be practical, since `mio` and the libraries and apps built on top of it need to interoperate with Rust's stdlib, which is in turn built on libc.

So it's not a "must-have", but it's already quite useful, and I expect it will become even more useful over time.  For example, we expect that WASI 0.2.x will have a `forward` call which tells the host to forward a `wasi:io/streams/output-stream` to a `wasi:io/streams/input-stream`, offloading that responsibility from the guest to the host and allowing it to proceed even after the guest is uninstantiated (optionally reinstantiating it if necessary to report failure).  Being able to e.g. forward the contents of a large file to an HTTP response body using that mechanism would be quite useful, and part of making it ergonomic is to be able to convert a libc file descriptor to a WASI `input-stream`.

Reply:
We should probably document that the WASI 0.2 ABI isn't stable for now. And it is worth thinking about whether we should make these table entries opaque to avoid exposing them in the ABI. That said, I do think we'll want to have public API for working with the table, to support interop between code that uses handles directly and code that uses libc fds.

Reply:
In the interest of keeping this whole effort moving forward, I'll go ahead and move this file to `headers/private` for now, and we can revisit later.